randovania/generator/hint_distributor.py

```diff
@@ -1,50 +1,87 @@
 from randovania.game_description.game_patches import GamePatches
 from randovania.game_description.hint import (
+    PRECISION_PAIR_UNASSIGNED,
     HintItemPrecision,
     HintLocationPrecision,
-    HintRelativeAreaName,
     JokeHint,
     LocationHint,
     PrecisionPair,
-    RelativeData,
-    RelativeDataArea,
-    RelativeDataItem,
     is_unassigned_location,
 )
 from randovania.game_description.resources.pickup_index import PickupIndex
 from randovania.generator.filler.filler_library import UnableToGenerate
-from randovania.generator.filler.player_state import PlayerState
-from randovania.lib import random_lib
+from randovania.generator.filler.player_state import HintState, PlayerState
 from randovania.resolver import debug
 
 if TYPE_CHECKING:
-    from randovania.game_description.db.area import Area
     from randovania.game_description.db.node_identifier import NodeIdentifier
     from randovania.game_description.db.region_list import RegionList
+    from randovania.game_description.hint_features import HintFeature, PickupHintFeature
     from randovania.game_description.pickup.pickup_entry import PickupEntry
     from randovania.generator.filler.filler_configuration import PlayerPool
     from randovania.generator.pre_fill_params import PreFillParams
 
 HintProvider = Callable[[PlayerState, GamePatches, Random, PickupIndex], LocationHint | None]
 
-
-def _not_empty(it: Iterator) -> bool:
-    return sum(1 for _ in it) > 0
+HintTargetPrecision = tuple[PickupIndex, PrecisionPair]
 
 
-HintTargetPrecision = tuple[PickupIndex, PrecisionPair]
+FeatureT = TypeVar("FeatureT")
+PrecisionT = TypeVar("PrecisionT")
+
+
+def choose_feature(
+    elements_with_feature: dict[FeatureT | PrecisionT, Collection[Any]],
+    total_elements: int,
+    detailed_precision: PrecisionT,
+    element_features: Collection[FeatureT],
+    rng: Random,
+    mean: float,
+    std_dev: float,
+) -> FeatureT | PrecisionT:
+    # arbitrarily increased until it felt good
+    DEGREE = 30
+    feature_precisions = {
+        feature: math.pow((total_elements - len(elements_with_feature[feature])) / (total_elements - 1), DEGREE)
+        for feature in elements_with_feature
+    }
+    feature_precisions = {
+        feature: ft_precision
+        for feature, ft_precision in feature_precisions.items()
+        if ft_precision < 1.0
+        # exclude any features that would only point to a single element
+    }
+    feature_precisions[detailed_precision] = 1.0
+    debug.debug_print(str(feature_precisions))
+
+    target_precision = rng.gauss(mean, std_dev)
+    target_precision = min(max(target_precision, 0.0), 1.0)
+    debug.debug_print(f"  * Target precision: {target_precision}")
+
+    possible_features: list[FeatureT | PrecisionT] = sorted(
+        feature for feature in element_features if feature in feature_precisions
+    )
+    possible_features.extend(feature for feature in feature_precisions if isinstance(feature, type(detailed_precision)))
```

Comment:
this looks pretty bad

Reply:
how do you mean?

Reply:
`isinstance(..., type(obj))` feels pretty weird